In [5]:
# Import the numpy and pandas package

import numpy as np
import pandas as pd
import


In [8]:
advertising = pd.read_csv("data_housing.csv")
advertising.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  Price  
0     15.3  396.90   4.98   24.0  
1     17.8  396.90   9.14   21.6  
2     17.8  392.83   4.03   34.7  
3     18.7  394.63   2.94   33.4  
4     18.7  396.90   5.33   36.2

# :Attribute Information (in order): 
# - CRIM per capita crime rate by town 
# - ZN proportion of residential land zoned for lots over 25,000 sq.ft. 
# - INDUS proportion of non-retail business acres per town 
# - CHAS Charles River dummy variable (= 1 if tract bounds river; 0 otherwise) 
# - NOX nitric oxides concentration (parts per 10 million)
# - RM average number of rooms per dwelling 
# - AGE proportion of owner-occupied units built prior to 1940 
# - DIS weighted distances to five Boston employment centres 
# - RAD index of accessibility to radial highways 
# - TAX full-value property-tax rate per $10,000 
# - PTRATIO pupil-teacher ratio by town 
# - B 1000(Bk - 0.63)^2 where Bk is the proportion of black people by town 
# - LSTAT % lower status of the population

Missing value check

In [9]:
advertising.isnull().mean()

CRIM       0.0
ZN         0.0
INDUS      0.0
CHAS       0.0
NOX        0.0
RM         0.0
AGE        0.0
DIS        0.0
RAD        0.0
TAX        0.0
PTRATIO    0.0
B          0.0
LSTAT      0.0
Price      0.0
dtype: float64

In [11]:
advertising.head(2)

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   

   PTRATIO      B  LSTAT  Price  
0     15.3  396.9   4.98   24.0  
1     17.8  396.9   9.14   21.6

In [14]:
advertising.nunique()

CRIM       504
ZN          26
INDUS       76
CHAS         2
NOX         81
RM         446
AGE        356
DIS        412
RAD          9
TAX         66
PTRATIO     46
B          357
LSTAT      455
Price      229
dtype: int64

In [17]:
num_cols=["CRIM","ZN","INDUS","NOX","RM","DIS","AGE","TAX","PTRATIO","B","LSTAT","Price"] 
cat_cols=["CHAS","RAD"]


## EDA

In [20]:
# univariate Analysis
for i in num_cols:
    sns.histplot(x=advertising[i])
    plt.show()

for i in cat_cols:
    sns.countplot(x=advertising[i])
    plt.show()
x    

IndentationError: expected an indented block after 'for' statement on line 6 (1070207725.py, line 6)

In [21]:
# Bivariate Analysis
# sns.pairplot(df(num_cols])
# plot.show()

for i in num_cols:
    sns.scatterplot(x=advertising[i],y=advertising["Price"])
    plt.show()

NameError: name 'sns' is not defined

In [22]:
# Multivariate Analysis
plt.figure(figsize=12,6)
sns.heatmap(advertising[num_cols],corr(),annot=True)
plt.show()
    

SyntaxError: unmatched ')' (4247537282.py, line 2)

    # Dummy Variable Creation one hot encoding

### Wherever we have more than 2 categories in categorical features, we need to create dummies for them

In [23]:
advertising[cat_cols].nunique()

CHAS    2
RAD     9
dtype: int64

In [27]:
# number of dumies =n-1
dum=pd.get_dummies(advertising["RAD"],drop_first=True,dtype=int)
dum.shape

(506, 8)

In [28]:
dum.head()


2.0   3.0   4.0   5.0   6.0   7.0   8.0   24.0
0     0     0     0     0     0     0     0     0
1     1     0     0     0     0     0     0     0
2     1     0     0     0     0     0     0     0
3     0     1     0     0     0     0     0     0
4     0     1     0     0     0     0     0     0

In [31]:
df1=pd.concat([advertising,dum],axis=1)
df1.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  ...  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0  ...   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0  ...   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0  ...   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0  ...   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0  ...   

   LSTAT  Price  2.0  3.0  4.0  5.0  6.0  7.0  8.0  24.0  
0   4.98   24.0    0    0    0    0    0    0    0     0  
1   9.14   21.6    1    0    0    0    0    0    0     0  
2   4.03   34.7    1    0    0    0    0    0    0     0  
3   2.94   33.4    0    1    0    0    0    0    0     0  
4   5.33   36.2    0    1    0    0    0    0    0     0  

[5 rows x 22 columns]

In [32]:
df1=df1.drop(["RAD"],axis=1)

In [34]:
df1.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS    TAX  PTRATIO  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  296.0     15.3   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  242.0     17.8   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  242.0     17.8   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  222.0     18.7   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  222.0     18.7   

   ...  LSTAT  Price  2.0  3.0  4.0  5.0  6.0  7.0  8.0  24.0  
0  ...   4.98   24.0    0    0    0    0    0    0    0     0  
1  ...   9.14   21.6    1    0    0    0    0    0    0     0  
2  ...   4.03   34.7    1    0    0    0    0    0    0     0  
3  ...   2.94   33.4    0    1    0    0    0    0    0     0  
4  ...   5.33   36.2    0    1    0    0    0    0    0     0  

[5 rows x 21 columns]

In [37]:
df1.shape

(506, 21)

In [39]:
df1.columns

Index([   'CRIM',      'ZN',   'INDUS',    'CHAS',     'NOX',      'RM',
           'AGE',     'DIS',     'TAX', 'PTRATIO',       'B',   'LSTAT',
         'Price',       2.0,       3.0,       4.0,       5.0,       6.0,
             7.0,       8.0,      24.0],
      dtype='object')

In [40]:
df1.columns=df1.columns.astype(str)

## Train-test split

In [50]:
X=df1.drop(["Price"],axis=1)
y=df1["Price"]

In [51]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.2,random_state=100)

In [52]:
print(X_train.shape)
print(X_test.shape)

(404, 20)
(102, 20)


## Scaling

In [56]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
X_train=scaler.fit_transform(X_train)
X_test=scaler.transform(X_test)
# as test data is unseen data we can`t fit on it


In [57]:
X_test

array([[-0.42553569,  2.98004576, -1.3851609 , ..., -0.19636805,
        -0.22215006, -0.58497238],
       [-0.37299822, -0.49577809, -0.70194846, ..., -0.19636805,
         4.50146175, -0.58497238],
       [-0.4245489 , -0.49577809,  0.13454882, ..., -0.19636805,
        -0.22215006, -0.58497238],
       ...,
       [ 0.2063882 , -0.49577809,  1.03527975, ..., -0.19636805,
        -0.22215006,  1.70948241],
       [-0.39261429, -0.49577809,  1.58856504, ..., -0.19636805,
        -0.22215006, -0.58497238],
       [ 0.52737117, -0.49577809,  1.03527975, ..., -0.19636805,
        -0.22215006,  1.70948241]])

In [59]:
X_train_df=pd.DataFrame(X_train,columns=X.columns)

X_test_df=pd.DataFrame(X_test,columns=X.columns)

In [61]:
X_train_df.head(2)

CRIM        ZN     INDUS      CHAS       NOX        RM       AGE  \
0  1.892434 -0.495778  1.035280 -0.288288  1.019549 -0.074765  1.128863   
1 -0.327672 -0.495778 -0.161802 -0.288288 -0.077387 -0.218185 -0.552953   

        DIS       TAX   PTRATIO         B     LSTAT       2.0       3.0  \
0 -1.175156  1.543468  0.813975  0.387891  1.262678 -0.203069 -0.293294   
1 -0.561852 -0.600514 -0.039554  0.424998 -0.910043 -0.203069 -0.293294   

        4.0       5.0       6.0       7.0      8.0      24.0  
0 -0.535373 -0.543021 -0.239983 -0.196368 -0.22215  1.709482  
1  1.867857 -0.543021 -0.239983 -0.196368 -0.22215 -0.584972

In [62]:
2# feature selection

2

In [64]:
from sklearn.feature_selection import RFE
from sklearn.linear_model import LinearRegression


In [67]:
estimator=LinearRegression()
selector=RFE(estimator,n_features_to_select=5)

In [68]:
selector=selector.fit(X_train_df,y_train)
selector.support_

array([False, False, False, False,  True,  True, False,  True, False,
        True, False,  True, False, False, False, False, False, False,
       False, False])

In [70]:
selected_features=X_train_df.columns[selector.support_]
print(selected_features)

Index(['NOX', 'RM', 'DIS', 'PTRATIO', 'LSTAT'], dtype='object')


In [71]:
X_train=X_train_df[selected_features]
X_test=X_test_df[selected_features]

In [73]:
import sklearn
sklearn.__version__

'1.3.2'

In [74]:
## Modelling

In [75]:
import statsmodels.api as sm

In [76]:
X_train_sm=sm.add_constant(X_train)
X_test_sm=sm.add_constant(X_test)

In [77]:
model1=sm.OLS(np.array(y_train),X_train_sm)
res1=model1.fit()
res1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.699
Model:                            OLS   Adj. R-squared:                  0.695
Method:                 Least Squares   F-statistic:                     185.1
Date:                Sun, 21 Jul 2024   Prob (F-statistic):          1.83e-101
Time:                        16:31:04   Log-Likelihood:                -1219.0
No. Observations:                 404   AIC:                             2450.
Df Residuals:                     398   BIC:                             2474.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         22.6210      0.248     91.260      0.000      22.134      23.108
NOX           -2.0015      0.418     -4.789      0.000      -2.823      -1.180
RM             2.8567      0.327      8.733      0.000       2.214       3.500
DIS           -2.3662      0.393     -6.019      0.000      -3.139      -1.593
PTRATIO       -2.1842      0.271     -8.073      0.000      -2.716      -1.652
LSTAT         -4.0517      0.383    -10.575      0.000      -4.805      -3.299
==============================================================================
Omnibus:                      143.828   Durbin-Watson:                   2.089
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              589.252
Skew:                           1.531   Prob(JB):                    1.11e-128
Kurtosis:                       8.063   Cond. No.                         3.53
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [78]:
# accepetable Range for p-value is <0.05
# Acceptable Range for VIF is <5

In [80]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

vif_data=pd.DataFrame()
vif_data["Feature"]=X_train_sm.columns
vif_data["VIF"]=[variance_inflation_factor(X_train_sm.values,i)
                 for i in range(len(X_train_sm.columns))]
vif_data

Feature       VIF
0    const  1.000000
1      NOX  2.842611
2       RM  1.741587
3      DIS  2.515516
4  PTRATIO  1.191260
5    LSTAT  2.389176

# Linear Regressioin Model (line of best fit)
y=2.0015*NOX+2.8567*RM-2.3662*DIS-2.1842*PTRATIO-4.0517*LSTAT+22.6210

In [82]:
X_train_sm.columns

Index(['const', 'NOX', 'RM', 'DIS', 'PTRATIO', 'LSTAT'], dtype='object')